### ISSUE TITLE
A question about pre-computed predictions of LAGO and fCLSWGAN

### ISSUE THREAD

**Issue by J4c0p0** [yuvalatzmon/COSMO#1]
Dear @yuvalatzmon,

first of all let me thank you for having shared the code of your CVPR'19 paper, which I found very interesting!

I was playing with your code, which works perfectly, but I have a few questions that I would like to humbly ask you. I am referring to the .npz files in which you stored the predictions of LAGO and fCLSWGAN (XianGAN) backbones.

1. To my understanding, in "_pred_gzsl_test.npz_" you concatenated the predictions carried out on test instances from both seen and unseen classes. Can you please tell me in which order you stacked the two groups? I suppose that you stored seen first and then unseen ones, but I would like to double-check this.

2. When loading "_pred_gzsl_val.npz_" from XianGAN, I am experiencing a mismatch in  the dimensions with the actual validation set proposed in [The Good, the Band and the Ugly of ZSL, Xian et al. CVPR 2017]. For instance, on AWA1, _val_loc_ indexes have 7926 entries, when I load _pred_gzsl_val.npz_ I am having a matrix with 11330 logits. I am right in saying that there should be a precise match in those dimensions?

Thank you very much in advance for your reply.

**Comment by yuvalatzmon** [comment]
Dear Jacopo,

Thank you for your interest in our paper!

About Q1:
The labels in `data['Y_GZSLtest']` correspond with the samples of
"pred_gzsl_test.npz". You can see in Line 112 of main_cosmo.py
<https://github.com/yuvalatzmon/COSMO/blob/master/src/main_cosmo.py#L112>,
that I first stack the samples of the seen classes and then the samples of
the unseen classes.
As another alternative, you can infer it from that labels by doing
`np.in1d(data['Y_GZSLtest'], data['seen_classes_test'])`

About Q2:
There are two sources of (legit) mismatch with val_loc:
(1) Our GZSLval split add samples of *seen* classes to the validation set,
which we take from the training set. The original split by Xian does not
has samples of seen classes.

(2) Our validation set uses 6372 samples of unseen-val classes
(`print(np.in1d(data['Y_GZSLval'], data['unseen_classes_val']).sum()`).
This is lower than 7926 samples (in Xian's `val_loc`),  because 1554
samples in `val_loc` overlap with the seen *test set*. Therefore, we take
them out in order to avoid overfitting. You can see it in line 254 in
load_xian_data.py
<https://github.com/yuvalatzmon/COSMO/blob/master/src/load_xian_data.py#L254>
<https://github.com/yuvalatzmon/COSMO/blob/master/src/load_xian_data.py#L254>

**Comment by J4c0p0** [comment]
Dear @yuvalatzmon,

thank you very much for your quick response!! I hope not to abuse your patience with another quick question that I would like to kindly ask you.

In practice, your validation set is the concatenation of two set of indexes: 1) the _val_loc_ (once you elimitated any overlap with _test_seen_loc_) and 2) the 2:1 split that you computed here <https://github.com/yuvalatzmon/COSMO/blob/a8875607c7dde60ad05a7e7ddb733a2a2af0168a/src/gating.py#L32> and for which you did not apply any shuffling. Am I correct in saying that?

Thank you again!

**Comment by yuvalatzmon** [comment]
Hi Jacopo,

Yes, I did not shuffle the validation set data.

Yuval

**Comment by J4c0p0** [comment]
Thank you very much, but I still see something unclear to me: on AWA1, if I take the _train_loc_ of Xian et al. CVPR 2017 and perform your recommended 2:1 subsampling, I get a 8432-dimensional vector. Then, I concatenate to it all indexes which are in _val_loc_ once the commonalities with _test_seen_loc_ are removed. As a result, I obtain a total validation set of 14804 elements. 

Unfortunately, I still see a mismatch with the shape of the 11330 logits that you provided in your _pred_gzsl_val.npz_ file (XianGAN, AWA1).

I am sorry to still bothering you, but I am really puzzled.
Thank you in advance for your reply.

**Comment by yuvalatzmon** [comment]
Hi Jacopo,

>>>if I take the *train_loc* of Xian et al. CVPR 2017 and perform your recommended 2:1 subsampling, I get a 8432-dimensional vector

That's not what we are doing. At the first phase of cross validation, we
take the seen train set (minus seen val set and seen test set)  to train
the experts. (see "X_seen_train" at link
<https://github.com/yuvalatzmon/COSMO/blob/master/src/load_xian_data.py#L326>
and
link2
<https://github.com/yuvalatzmon/COSMO/blob/master/src/cosmo_misc.py#L464>)
The 2:1 sub sampling is only used to train the "gating" model. For that we
sub sample the "GZSLval" *validation* set, since it helps us to simulate
response of the experts to "unseen" classes.

I apologize for my slow responsiveness,  I am before a tight deadline and
my bandwidth is limited.

**Comment by J4c0p0** [comment]
Thanks again for your kind reply, I tried to go back in your code exploiting the links you provided, I appreciated it! Unfortunately, it is still unclear to which data do the 11330 logits in _/COSMO/data/XianGAN_predictions/AWA1/pred_gzsl_val.npz_  refer to.

**Comment by yuvalatzmon** [comment]
Hi Jacopo,
I apologize again for the delayed reply. I'll try to clarify better.

`train_loc` has 16864 samples, which overlaps with `test_seen_loc` by 3404
samples. This leaves us with 13460 samples that we can use for training /
validation.
Of those samples, we sample 4958 for the *seen* validation set. We join
these samples with the 6372 *unseen* validation samples (that we discussed
earlier) and get a total of 4958+6372=11330 GZSLval samples.

As a side note: Taking 4958 samples for validation leaves us with
13460-4958=8502 samples for training the first phase of cross validation.

I hope it is helpful
Best!
Yuval

**Comment by J4c0p0** [comment]
Thank you very much Yuval (and sorry for my late reply too). I think that now things are clear. Thanks for your time!
